File: app/src/main/java/ch/berta/fabio/popularmovies/data/storage/MovieContract.java
/*
 * Copyright (c) 2015 Fabio Berta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.berta.fabio.popularmovies.data.storage;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.net.Uri;
import android.provider.BaseColumns;

/**
 * Created by fabio on 12.10.15.
 */
public class MovieContract {

    public static final String CONTENT_AUTHORITY = "ch.berta.fabio.popularmovies";
    public static final Uri BASE_CONTENT_URI = Uri.parse("content://" + CONTENT_AUTHORITY);

    public static final String PATH_MOVIE = "movie";
    public static final String PATH_REVIEW = "review";
    public static final String PATH_VIDEO = "video";

    public static class Movie implements BaseColumns {

        public static final Uri CONTENT_URI =
                BASE_CONTENT_URI.buildUpon().appendPath(PATH_MOVIE).build();
        public static final String CONTENT_TYPE =
                ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_MOVIE;
        public static final String CONTENT_ITEM_TYPE =
                ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_MOVIE;

        public static final String TABLE_NAME = "movie";

        public static final String COLUMN_DB_ID = "db_id";
        public static final String COLUMN_TITLE = "title";
        public static final String COLUMN_RELEASE_DATE = "release_date";
        public static final String COLUMN_VOTE_AVERAGE = "vote_average";
        public static final String COLUMN_PLOT = "plot";
        public static final String COLUMN_POSTER = "poster";
        public static final String COLUMN_BACKDROP = "backdrop";

        public static final String SORT_DEFAULT = TABLE_NAME + "." + _ID + " DESC";
        public static final String SORT_BY_RELEASE_DATE =  TABLE_NAME + "." + COLUMN_RELEASE_DATE + " DESC";

        public static Uri buildMovieUri(long id) {
            return ContentUris.withAppendedId(CONTENT_URI, id);
        }

        public static Uri buildMovieByDbIdUri(int dbId) {
            return CONTENT_URI.buildUpon()
                    .appendPath(COLUMN_DB_ID)
                    .appendPath(String.valueOf(dbId))
                    .build();
        }

        public static Uri buildMovieWithReviewsAndTrailersUri(long id) {
            return CONTENT_URI.buildUpon()
                    .appendPath(PATH_REVIEW)
                    .appendPath(PATH_VIDEO)
                    .appendPath(String.valueOf(id))
                    .build();
        }

        public static String getMovieIdFromUri(Uri uri) {
            return uri.getPathSegments().get(3);
        }
    }

    public static class Review implements BaseColumns {
        public static final Uri CONTENT_URI =
                BASE_CONTENT_URI.buildUpon().appendPath(PATH_REVIEW).build();
        public static final String CONTENT_TYPE =
                ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_REVIEW;
        public static final String CONTENT_ITEM_TYPE =
                ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_REVIEW;

        public static final String TABLE_NAME = "review";

        public static final String COLUMN_MOVIE_ID = "movie_id";
        public static final String COLUMN_AUTHOR = "author";
        public static final String COLUMN_CONTENT = "content";
        public static final String COLUMN_URL = "url";

        public static final String INDEX_MOVIE_ID = TABLE_NAME + "_" + COLUMN_MOVIE_ID + "_idx";

        public static Uri buildReviewUri(long id) {
            return ContentUris.withAppendedId(CONTENT_URI, id);
        }

        public static String getMovieIdFromUri(Uri uri) {
            return uri.getPathSegments().get(2);
        }
    }

    public static class Video implements BaseColumns {
        public static final Uri CONTENT_URI =
                BASE_CONTENT_URI.buildUpon().appendPath(PATH_VIDEO).build();
        public static final String CONTENT_TYPE =
                ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_VIDEO;
        public static final String CONTENT_ITEM_TYPE =
                ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + CONTENT_AUTHORITY + "/" + PATH_VIDEO;

        public static final String TABLE_NAME = "video";

        public static final String COLUMN_MOVIE_ID = "movie_id";
        public static final String COLUMN_NAME = "name";
        public static final String COLUMN_KEY = "key";
        public static final String COLUMN_SITE = "site";
        public static final String COLUMN_SIZE = "size";
        public static final String COLUMN_TYPE = "type";

        public static final String INDEX_MOVIE_ID = TABLE_NAME + "_" + COLUMN_MOVIE_ID + "_idx";

        public static Uri buildVideoUri(long id) {
            return ContentUris.withAppendedId(CONTENT_URI, id);
        }

        public static String getMovieIdFromUri(Uri uri) {
            return uri.getPathSegments().get(2);
        }
    }
}

File: app/src/main/java/ch/berta/fabio/popularmovies/data/storage/MovieContract.kt
/*
 * Copyright (c) 2015 Fabio Berta
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.berta.fabio.popularmovies.data.storage

import android.content.ContentResolver.CURSOR_DIR_BASE_TYPE
import android.content.ContentResolver.CURSOR_ITEM_BASE_TYPE
import android.content.ContentUris
import android.net.Uri
import android.provider.BaseColumns
import ch.berta.fabio.popularmovies.BuildConfig

/**
 * Defines the structure of the local sqlite database.
 */
object MovieContract {

    const val CONTENT_AUTHORITY = BuildConfig.APPLICATION_ID
    const val PATH_MOVIE = "movie"
    const val PATH_REVIEW = "review"
    const val PATH_VIDEO = "video"

    val baseContentUri: Uri = Uri.parse("content://" + CONTENT_AUTHORITY)

    object Movie : BaseColumns {
        const val CONTENT_TYPE = "$CURSOR_DIR_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_MOVIE"
        const val CONTENT_ITEM_TYPE = "$CURSOR_ITEM_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_MOVIE"
        const val TABLE_NAME = "movie"

        const val COLUMN_DB_ID = "db_id"
        const val COLUMN_TITLE = "title"
        const val COLUMN_RELEASE_DATE = "release_date"
        const val COLUMN_VOTE_AVERAGE = "vote_average"
        const val COLUMN_PLOT = "plot"
        const val COLUMN_POSTER = "poster"
        const val COLUMN_BACKDROP = "backdrop"

        const val SORT_DEFAULT = "$TABLE_NAME.${BaseColumns._ID} DESC"
        const val SORT_BY_RELEASE_DATE = "$TABLE_NAME.$COLUMN_RELEASE_DATE DESC"

        val contentUri: Uri = baseContentUri.buildUpon().appendPath(PATH_MOVIE).build()

        fun buildMovieUri(id: Long): Uri {
            return ContentUris.withAppendedId(contentUri, id)
        }

        fun buildMovieByDbIdUri(dbId: Int): Uri {
            return contentUri.buildUpon()
                    .appendPath(COLUMN_DB_ID)
                    .appendPath(dbId.toString())
                    .build()
        }

        fun buildMovieWithReviewsAndTrailersUri(id: Long): Uri {
            return contentUri.buildUpon()
                    .appendPath(PATH_REVIEW)
                    .appendPath(PATH_VIDEO)
                    .appendPath(id.toString())
                    .build()
        }

        fun getMovieIdFromUri(uri: Uri): String {
            return uri.pathSegments[3]
        }
    }

    object Review : BaseColumns {
        const val CONTENT_TYPE = "$CURSOR_DIR_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_REVIEW"
        const val CONTENT_ITEM_TYPE = "$CURSOR_ITEM_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_REVIEW"
        const val TABLE_NAME = "review"

        const val COLUMN_MOVIE_ID = "movie_id"
        const val COLUMN_AUTHOR = "author"
        const val COLUMN_CONTENT = "content"
        const val COLUMN_URL = "url"
        const val INDEX_MOVIE_ID = TABLE_NAME + "_" + COLUMN_MOVIE_ID + "_idx"

        val contentUri: Uri = baseContentUri.buildUpon().appendPath(PATH_REVIEW).build()

        fun buildReviewUri(id: Long): Uri {
            return ContentUris.withAppendedId(contentUri, id)
        }

        fun buildReviewsFromMovieUri(movieId: Long): Uri {
            return contentUri.buildUpon().appendPath(PATH_MOVIE).appendPath(
                    movieId.toString()).build()
        }

        fun getMovieIdFromUri(uri: Uri): String {
            return uri.pathSegments[2]
        }
    }

    object Video : BaseColumns {
        const val CONTENT_TYPE = "$CURSOR_DIR_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_VIDEO"
        const val CONTENT_ITEM_TYPE = "$CURSOR_ITEM_BASE_TYPE/$CONTENT_AUTHORITY/$PATH_VIDEO"
        const val TABLE_NAME = "video"

        const val COLUMN_MOVIE_ID = "movie_id"
        const val COLUMN_NAME = "name"
        const val COLUMN_KEY = "key"
        const val COLUMN_SITE = "site"
        const val COLUMN_SIZE = "size"
        const val COLUMN_TYPE = "type"
        const val INDEX_MOVIE_ID = TABLE_NAME + "_" + COLUMN_MOVIE_ID + "_idx"

        val contentUri: Uri = baseContentUri.buildUpon().appendPath(PATH_VIDEO).build()

        fun buildVideoUri(id: Long): Uri {
            return ContentUris.withAppendedId(contentUri, id)
        }

        fun buildVideosFromMovieUri(movieId: Long): Uri {
            return contentUri.buildUpon().appendPath(PATH_MOVIE).appendPath(
                    movieId.toString()).build()
        }

        fun getMovieIdFromUri(uri: Uri): String {
            return uri.pathSegments[2]
        }
    }
}
